### ISSUE TITLE
[Runtime crash][Generics] Setting closure property of generic struct through generic function cause crash in runtime

### ISSUE THREAD

**Issue by Alkenso** [apple/swift#65178]
**Description**
Given structure with specified generic type T `struct Foo<T>`.
The structure has property of closure type that uses this generic type `var action: ((T) -> Void)?`.
Setting this property through generic function using WritableKeyPath causes crash in runtime

The issue is reproduces on both latest Swift (5.8 in Xcode 14.3) and older one (5.7.2 in Xcode 14.2).

**Steps to reproduce**
```
struct Foo<T> {
    var action: ((T) -> Void)?
}

extension Foo {
    func set<U>(_ keyPath: WritableKeyPath<Self, U>, _ value: U?) -> Self {
        guard let value else { return self }
        var copy = self
        copy[keyPath: keyPath] = value
        return copy
    }
}

func willCrash() {
    _ = Foo<Int>().set(\.action) { _ in }
    // crashes with BAD_ACCESS on function exit.
}

willCrash()
```

**Expected behavior**
The code does not cause the crash

**Environment**
- Swift compiler version info ` 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)`
- Xcode version info `Xcode 14.3 Build version 14E222b`
- Deployment target: does not matter

**Comment by Alkenso** [comment]
cc @asl , @asavonic

**Comment by Alkenso** [comment]
fixed in 5.9
